### ISSUE TITLE
Add Settings.renderTable

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#8893]
Some users might want to avoid table qualification of columns where this is not strictly necessary. For example:

```sql
-- This
SELECT a, b, c FROM t;

-- Instead of this
SELECT t.a, t.b, t.c FROM t;
```

This would work in addition to the omitted schema qualification:

```sql
SELECT s.t.a, s.t.b, s.t.c FROM s.t;
```

There are numerous ways such column names can be ambiguous. Hence, we offer the following flag values:

- `ALWAYS` (always qualify columns with a table. The default and current behaviour)
- `NEVER` (never qualify columns with a table. This risks breaking a lot of queries)
- `WHEN_MULTIPLE_TABLES` (qualify columns only when there are mutliple tables in the `FROM` clause)
- `WHEN_AMBIGUOUS_COLUMNS` (qualify columns only when we *know* there is an ambiguity. This requires a bit of calculation when generating the SQL string. It is currently unclear whether this is worth implementing for general purpose jOOQ, but it could definitely be useful for the parser).

A special use-case is the parser when it uses meta lookups (#7163), which currently produce qualified column references from what the parser sees as unqualified column references.

**Comment by lukaseder** [comment]
A previous version of the description mentioned these options:

- `ALWAYS` (always qualify columns with a table. The default and current behaviour)
- `NEVER` (never qualify columns with a table. This risks breaking a lot of queries)
- `WHEN_NEEDED` (qualify columns only when we *know* there is an ambiguity. This requires a bit of calculation when generating the SQL string. It is currently unclear whether this is worth implementing for general purpose jOOQ, but it could definitely be useful for the parser).

We will only implement a boolean value that will never qualify columns with their table name, irrespective of possible ambiguities.

**Comment by internetstaff** [comment]
It'd be great if `WHEN_NEEDED` was possible. Our particular issue is the Jooq-generated SQL is extremely large and somewhat unreadable for audit logging.

**Comment by lukaseder** [comment]
> It'd be great if `WHEN_NEEDED` was possible. Our particular issue is the Jooq-generated SQL is extremely large and somewhat unreadable for audit logging.

Because of rendered tables? I mean, you can already remove the schema using `Settings.renderSchema`, then remove the quoting, and use aliases in your code to make it smaller...

**Comment by internetstaff** [comment]
We've removed schemas and quoting, but we have a lot of long table names with lots of columns! :)

You're right on aliasing, we can employ that - it's just a bit of effort compared to an automated solution (if practical)

Thanks!

**Comment by lukaseder** [comment]
Well, you could enhance the code generator and generate aliased versions of all the tables instead of the unaliased ones, i.e. override the logic that generates the `Tables.java` class...

**Comment by lukaseder** [comment]
That would even be transparent to your existing code! You could use a 4-letter table abbreviation convention for that: https://blog.jooq.org/2019/10/29/a-guide-to-sql-naming-conventions/

**Comment by npostnikova** [comment]
Hi there,

Is there a way to get rid of table qualification at the moment?

I'm generating update statements using `SQLDialect.H2`. Also, I need to execute their string representation using postgres, which is unhappy with qualification.

Thanks!

**Comment by lukaseder** [comment]
Thanks @npostnikova. Why do you want to generate the statement with the H2 dialect when you will run it with PostgreSQL? I'm sure your *actual* question is something else, and this feature request here will not sufficiently solve this (and other) problems that you may have. Maybe, better state your full problem description in a new support request? https://github.com/jOOQ/jOOQ/issues/new/choose

**Comment by npostnikova** [comment]
Well, in our case, the client has H2. It generates the request and executes it, but additionally it sends queries to the server, which has PostgreSQL.
Update statements turned out to be incompatible with PostgreSQL, something like `update sample set sample.name = 'name' where sample.id = 666` is generated, and `sample.name` is incorrect for postgres.

I guess this case is quite weird, we should redesign the protocol, not you. Just was curious whether there were some updates on this issue.
For now, the problem was solved by using `SQLDialect.POSTGRES`, which hasn't raised issues on the H2 side yet :)

**Comment by lukaseder** [comment]
That sounds creative :) (I'm assuming SQLi isn't a risk in your architecture?) Well, you could parse the SQL string with jOOQ's parser, and render it again with the PostgreSQL dialect...

Or better, render it twice on the client. Once with H2 and once with PostgreSQL, and send only the latter to the server.

As I hinted, there will be many other problems of subtle and not so subtle differences between dialects. The SQL generated by jOOQ is very much tailored towards a specific dialect (and version), so it's often unlikely you'll get portable SQL out of jOOQ. The column name qualification in update statements is only one such difference.

**Comment by lukaseder** [comment]
Regarding that:

> Just was curious whether there were some updates on this issue.

It's quite the low priority issue, and it wouldn't work well for you, because you'd stop qualifying columns *everywhere*, which would break many queries that have joins in them.

**Comment by npostnikova** [comment]
Yeah, I thought about rendering it twice as well, maybe we'll end up choosing the option :)

I'll keep that in mind, thanks for your help!

> because you'd stop qualifying columns everywhere

Oh, I see, sounds bad.

**Comment by trydofor** [comment]
> Hi there,
> 
> Is there a way to get rid of table qualification at the moment?
> 
> I'm generating update statements using `SQLDialect.H2`. Also, I need to execute their string representation using postgres, which is unhappy with qualification.
> 
> Thanks!

use DefaultVisitListener to render sql

```java
@Override
    public void visitStart(VisitContext context) {
        // only rendering
        if(context.renderContext() == null) return;

        QueryPart qp = context.queryPart();
        if (qp instanceof TableField) {
            TableField<?, ?> field = (TableField<?, ?>) qp;
            if (notAlias(field.getTable(), context.context()) == 0) {
                context.queryPart(DSL.field(field.getUnqualifiedName(), field.getDataType()));
            }
        }
        else if (qp instanceof QualifiedAsterisk) {
            QualifiedAsterisk asterisk = (QualifiedAsterisk) qp;
            if (notAlias(asterisk.qualifier(), context.context()) == 0) {
                context.queryPart(DSL.sql("*"));
            }
        }
    }
```
full code in 
https://github.com/trydofor/pro.fessional.wings/blob/master/wings-project/wings-faceless-jooq/src/main/java/pro/fessional/wings/faceless/database/jooq/listener/AutoQualifyFieldListener.java

**Comment by lukaseder** [comment]
I can see that a few folks have shared valid use-cases for this flag in the past, mostly everywhere a human needs to read things.

The reason why this was delayed was:

- `NEVER` is quite a harsh setting, and shouldn't really be used, though we can add it because it's simple to implement, and there might be a use-case I'm overlooking
- `WHEN_NEEDED` (original naming, now called `WHEN_AMBIGUOUS_COLUMNS`) is costly to compute

But there's an additional flag, which should be simple to implement:

- `WHEN_MULTIPLE_TABLES`. As soon as there are multiple tables in the `FROM` clause, it is "risky" to leave column names unqualified. But if there's only a single table for simple queries, then there's no risk of ambiguities.

Caveats as always:

- Plain SQL. In the presence of plain SQL tables, we don't know if there are ambiguities, so both `WHEN_NEEDED` and `WHEN_MULTIPLE_TABLES` should lead to full qualification. Only `NEVER` would turn that off, in those cases.

**Comment by lukaseder** [comment]
Implemented in jOOQ 3.18.0.

**Comment by lukaseder** [comment]
Some known issues include:

- https://github.com/jOOQ/jOOQ/issues/15993
- #15996 
- #15997
